### DISCUSSION TITLE
Is Python >= 3.12 OK for next release?

### DISCUSSION THREAD

**Discussion by MestreLion** [MestreLion/mcworldlib#20] [Polls]
This month I'll release a new MCWorldLib version, aggregating all the new features and contributions from the last 2 years. 

I'll also take the time to add some pending PRs, change the build system, upload to PyPI, and.... bump the Python version requirement from the old 3.8 to a more modern one, at least 3.10, possibly 3.11 and maybe even 3.12

My question is: Is Python 3.12 too new for a required minimum? I have no idea about the user base, hence a poll.

I know a library should strive to be compatible with the widest possible range, so preferably it should have its requirements as low as possible. Python 3.10 brings _many_ improvements to `typing`, and 3.11 has the godsend `typing.Self`. But for those there is `typing_extensions`.

Now 3.12 has the new `type` parameter syntax, including aliases as `type X = Y` and generics such as `def X[T](args: list[T]) -> T:`  without the clumsy `_T = typing.VarType("_T", bound=...)`, actually without even _importing_ `typing`. And MCWorldLib uses both aliases and generics _a lot_. It would at the same time greatly simplify MCWorldLib's typing annotations and make them more accurate, and a base library should strive for that too.

Ubuntu 24.04 has Python 3.12 as the default, but the still widely used 22.04 is only 3.10. One could always install the DeadSnakes PPA or use `pyenv`, but I know either approach is an extra hassle for the user.

**TL/DR: _What is YOUR highest _installed_, readily available Python?_ In other words:  Up to which Python version can a project require so you can `pip/pipx`-install it _right now_ without any extra steps in your main development system?**

**Comment by MestreLion** [comment]
Pinging all collaborators and seemingly interested people, judging by PR requests, issues, comments and reactions. I apologize in advance if this was uncalled for.

@ievans3024 , @misode , @NielsPichon , @Starrah , @atmosphericbeats , @fqhd , @Fhoughton , @17183248569 , @NPBiceps , @cullen6995 , @jason-green-io , @suzakuwcx , @9NONAGON , @ch-yx

**Comment by NielsPichon** [comment]
A few considerations:
* anaconda python, one of the major python providers on windows, supports up to python 3.12 if I am not mistaken. So no issue there. 
* Ubuntu LTS 24.04 also ships with 3.12.
* Many libraries haven't made the switch just yet however. I, for one, mostly use 3.10 and sometimes 3.11. 

Moving to 3.12 would be reasonable. And keeping up to date is good practice (even though my own libraries that depend on this project are kinda stuck in the past 😅 )

**Reply by MestreLion** [reply]
Thanks for the feedback!!!

So, if I got it right, while your _system default_ Python is 3.10, you **do** have 3.11 (but not 3.12) installed, correct? I mean, could you `pip`-install a project requiring Python 3.11 _right now_ in your system? Or would you have to, say, add the DeadSnakes PPA and `sudo apt install python3.11`  (or the equivalent in Anaconda) first?

I've updated the Poll to clarify what I'm actually asking. I want update this project's required python as high as I can (up to 3.12) _without_ creating a hassle to most users (whoever they might be, I have no idea! 😅).

Since updating a poll discards all votes, may I ask you to re-cast yours (to 3.11 or 3.10 depending on what's installed)? Truly sorry for the inconvenience!

**Reply by NielsPichon** [reply]
My project which depends on this library is currently locked on 3.11 yes. But I could consider updating this is not much of a hassle on my end.

I will cast my vote again.

**Comment by suzakuwcx** [comment]
Although my system's Python version is 3.12, I still recommend using Python 3.10 for the following reasons:

- Dependency Management Challenges:
    Python's dependency management can often be messy. For small system scripts, I typically stick to lightweight and widely-used libraries like requests. However, for larger projects, I rely on tools like micromamba or venv to manage dependencies and freeze their versions upfront. This approach minimizes compatibility issues caused by unexpected updates. Additionally, if a dependency error occurs, fixing it without side effects is almost impossible, often leaving behind a clutter of unnecessary files in ~/.local/.

- Library Compatibility:
    Many libraries have yet to support Python 3.12, which makes it less reliable for most projects.

To improve clarity, including a small installation example in the code section could be helpful.

**Reply by MestreLion** [reply]
Very solid points, thanks for your feedback!

> However, for larger projects, I rely on tools like micromamba or venv to manage dependencies and freeze their versions upfront. This approach minimizes compatibility issues caused by unexpected updates. Additionally, if a dependency error occurs, fixing it without side effects is almost impossible, often leaving behind a clutter of unnecessary files in ~/.local/.

Yes, installing any project in a Virtual Environment is not only the current best practice but almost mandatory to avoid headaches. I used to have very few, "essential" (and widely-compatible) libs and execs like `requests` and `black`  in my `~/.local/lib/python3.*/site-packages`, but currently I have **none**. Every single package is installed either in a per-project `venv` (if it's a lib that project depends on) or via `pipx` (for executables such as `black`, `mypy`, etc). I don't even have a `~/.local/lib` folder anymore!

> Many libraries have yet to support Python 3.12, which makes it less reliable for most projects.

Luckily MCWorldLib have only 3 dependencies: `numpy` and `tqdm`, both supporting latest Python *3.13*, and `nbtlib`, which while not very active still works great on 3.12 and latest Numpy.

> To improve clarity, including a small installation example in the code section could be helpful.

Yeah, great point. I'm surprised to notice the README lacks even a basic "Install" section. I'll add one in this release, something like this:

> Installing in a Virtual Environment is highly recommended, and the instructions below do so.
> Debian/Ubuntu users might have to install Python's `venv` module:
> ```
> sudo apt install python3-venv
> ```
> Run in your project/script root directory:
> ```sh
> python3 -m venv venv               # Create a `venv` Virtual Environment directory
> ./venv/bin/pip install mcworldlib  # Install MCWorldLib and its dependencies in the Virtual Environment
> ```

So, assuming installs in a venv (and the above README instructions), would 3.12 still be too harsh to require for users?

**Comment by MestreLion** [comment]
I'm currently inclined to, in this release, do only a very modest bump to Python 3.10, and use 3.11 features from `typing_extensions` (they're all `typing`-related), and wait an year or two to use 3.12's amazing revamped `type` syntax for aliases and TypeVars. Specially the `util.*Pos*` classes and my `nbt` wrapper will benefit a **lot** from it (and they're currently quite messy), but it's been working so far (in 3.6!), so I believe a more conservative approach is advisable.

In any case, I'll wait a week or so for more feedback before making a final decision.
